jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/FilesTask.java

```diff
@@ -0,0 +1,140 @@
+    }
+
+    // Print build.gradle
+    System.out.println(project.getBuildFile());
+
+    // Print settings.gradle
+    Path settingsFile = project.getProjectDir().toPath().resolve(Settings.DEFAULT_SETTINGS_FILE);
+    if (Files.exists(settingsFile)) {
+      System.out.println(settingsFile);
+    }
+
+    // Print sources
+    mainSourceSet.getAllJava().getSourceDirectories().forEach(System.out::println);
+
+    // Print resources
+    mainSourceSet.getResources().getSourceDirectories().forEach(System.out::println);
+
+    // Find project dependencies
+    for (Configuration configuration :
+        project.getConfigurations().getByName("runtime").getHierarchy()) {
+      for (Dependency dependency : configuration.getDependencies()) {
+        if (dependency instanceof ProjectDependency) {
+          // Keep track of project dependency jars
+          ProjectDependency projectDependency = (ProjectDependency) dependency;
+          String configurationName = projectDependency.getTargetConfiguration();
+          if (configurationName == null) {
+            configurationName = "default";
+          }
+          Project dependencyProject = projectDependency.getDependencyProject();
+          for (Configuration targetConfiguration :
+              dependencyProject.getConfigurations().getByName(configurationName).getHierarchy()) {
+            for (PublishArtifact artifact : targetConfiguration.getArtifacts()) {
+              projectDependencyJars.add(artifact.getFile());
+            }
+          }
+
+          // Print project dependency's build/source/resource files
+          listFilesForProject(dependencyProject, projectDependencyJars);
+        }
+      }
+    }
+  }
+
+  @Nullable private JibExtension jibExtension;
+
+  public FilesTask setJibExtension(JibExtension jibExtension) {
+    this.jibExtension = jibExtension;
+    return this;
+  }
+
+  @TaskAction
+  public void listFiles() {
+    Preconditions.checkNotNull(jibExtension);
+    Project project = getProject();
+
+    // If this is not the root project, print the root project's build.gradle and settings.gradle
+    if (project != project.getRootProject()) {
+      System.out.println(project.getRootProject().getBuildFile());
+      Path settingsFiles = project.getRootDir().toPath().resolve(Settings.DEFAULT_SETTINGS_FILE);
+      if (Files.exists(settingsFiles)) {
+        System.out.println(settingsFiles);
+      }
+    }
+
+    // Print extra layer
+    if (project.getPlugins().hasPlugin(JibPlugin.class)) {
+      System.out.println(jibExtension.getExtraDirectoryPath());
+    }
+
+    // Print sub-project sources
+    Set<File> skippedJars = new HashSet<>();
+    listFilesForProject(project, skippedJars);
+
+    // Print out SNAPSHOT non-project dependency jars
+    for (File file : project.getConfigurations().getByName("runtime")) {
+      if (!skippedJars.contains(file) && file.toString().contains("SNAPSHOT")) {
+        System.out.println(file);
+        skippedJars.add(file);
```

Comment:
It's a bit hard to follow the logic behind using the `Set<File> skippedJars` here. It might be good to isolate the mutation of `skippedJar` to a more encapsulated location, since `skippedJars` seems to be mutated here, and inside `listFilesForProject`. Also, in `listFilesForProject`, the set is called `projectDependencyJars`, but here it is called `skippedJars`. A comment would help explain the discretion in naming.

Reply:
The reason it's called `skippedJars` here and `projectDependencyJars` within the method is because within the method, the set is only used to keep track of project dependency jars, while here the set is also used to avoid printing the same files twice. I'll rethink this or add a comment.